File: sol/solution/0101-0200/0107/Solution.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  List<List<int>> levelOrderBottom(TreeNode? root) {
    List<List<int>> result = [];
    if (root == null) return result;
    List<TreeNode> queue = [root];
    while (queue.isNotEmpty) {
      int levelSize = queue.length;
      List<int> level = [];
      for (int i = 0; i < levelSize; i++) {
        TreeNode node = queue.removeAt(0);
        level.add(node.val);
        if (node.left != null) queue.add(node.left!);
        if (node.right != null) queue.add(node.right!);
      }
      result.add(level);
    }
    return result.reversed.toList();
  }
}

File: sol/solution/0101-0200/0107/Solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def levelOrderBottom(root: TreeNode): List[List[Int]] = {
        var result = List[List[Int]]()
        if (root == null) return result
        var queue = scala.collection.mutable.Queue[TreeNode](root)
        while (queue.nonEmpty) {
            val levelSize = queue.length
            var level = List[Int]()
            for (_ <- 0 until levelSize) {
                val node = queue.dequeue()
                level = level :+ node.value
                if (node.left != null) queue.enqueue(node.left)
                if (node.right != null) queue.enqueue(node.right)
            }
            result = level :: result
        } 
        result
    }
}